File: appinventor/components/src/com/google/appinventor/components/common/GyroSensorMode.java
// -*- mode: java; c-basic-offset: 2; -*-
// Copyright 2020 MIT, All rights reserved
// Released under the Apache License, Version 2.0
// http://www.apache.org/licenses/LICENSE-2.0

package com.google.appinventor.components.common;

import java.util.HashMap;
import java.util.Map;

/**
 * Defines a GyroSensorMode type used by the Lego Ev3 gyro sensor.
 */
public enum GyroSensorMode implements OptionList<String> {
  Angle("angle", 0),
  Rate("rate", 1);

  private final String value;
  private final int intValue;

  GyroSensorMode(String mode, int intMode) {
    this.value = mode;
    this.intValue = intMode;
  }

  public String toUnderlyingValue() {
    return value;
  }

  public Integer toInt() {
    return this.intValue;
  }

  private static final Map<String, GyroSensorMode> lookup = new HashMap<>();

  static {
    for (GyroSensorMode mode : GyroSensorMode.values()) {
      lookup.put(mode.toUnderlyingValue(), mode);
    }
  }

  public static GyroSensorMode fromUnderlyingValue(String mode) {
    return lookup.get(mode);
  }
}

File: appinventor/components-ios/src/GyroSensorMode.swift
// -*- mode: java; c-basic-offset: 2; -*-
// Copyright © 2025 MIT, All rights reserved
// Released under the Apache License, Version 2.0
// https://www.apache.org/licenses/LICENSE-2.0

@objc public class GyroSensorMode: NSObject, OptionList {
  @objc public static let Angle = GyroSensorMode("angle", 0)
  @objc public static let Rate = GyroSensorMode("rate", 1)

  private static let LOOKUP: [String: GyroSensorMode] = generateOptionsLookup(Angle, Rate)

  let value: String
  let intValue: Int32

  @objc private init(_ value: String, _ intValue: Int32) {
    self.value = value
    self.intValue = intValue
  }

  @objc public func toUnderlyingValue() -> AnyObject {
    return value as AnyObject
  }

  @objc public func toInt() -> Int32 {
    return intValue
  }

  @objc public static func fromUnderlyingValue(_ value: String) -> GyroSensorMode? {
    return LOOKUP[value]
  }
}
